p2p/server/server.go

```diff
@@ -147,26 +228,51 @@ func (s *Server) streamHandler(stream network.Stream) {
 	}
 }
 
+func (s *Server) trackClientLatency(start time.Time, err error) {
+	if s.metrics != nil {
+		if err != nil {
+			s.metrics.clientLatencyFailure.Observe(time.Since(start).Seconds())
+		} else {
+			s.metrics.clientLatency.Observe(time.Since(start).Seconds())
+		}
+	}
+}
+
 // Request sends a binary request to the peer. Request is executed in the background, one of the callbacks
 // is guaranteed to be called on success/error.
-func (s *Server) Request(ctx context.Context, pid peer.ID, req []byte, resp func([]byte), failure func(error)) error {
+func (s *Server) Request(
+	ctx context.Context,
+	pid peer.ID,
+	req []byte,
+	resp func([]byte),
+	failure func(error),
+) error {
```

Comment:
I just noticed `Request` isn't blocking at the moment. That seems incorrect to me. I would assume that I can discard the `ctx` after a function that receives it returns. Imagine the following code:

```go
func myFunc(ctx context.Context, ...) error {
	ctx, cancel := context.WithTimeout(context.Background(), 5 *time.Second)
	defer cancel()
	return srv.Request(ctx, ...)
```

This will aways fail immediately with `context.Canceled` and not after 5 seconds with `context.DeadlineExceeded` what I would expect.

Reply:
Is there an issue tracking the need to change this component? Holding on to parameters after a function returns can lead to problems and data races which are to debug later.

Reply:
there is 1st point in this issue https://github.com/spacemeshos/pm/issues/272